### DISCUSSION TITLE
JSON library

### DISCUSSION THREAD

**Discussion by metalefty** [neutrinolabs/xrdp#2004] [General]
I remember I had a talk with @matt335672 that in the context of sesman. It was a suggestion to switch communication between xrdp to sesman to JSON based new protocol. Do you remember it? 

So, what is the best/better JSON library do you think?

* [picojson](https://github.com/kazuho/picojson)
* [jansson](https://github.com/akheron/jansson)
* [parson](https://github.com/kgabis/parson)

**Comment by matt335672** [comment]
I do indeed remember that. It was in a gitter conversation.

I'd say at a minimum we'd need something which has a C interface, and is supported by FreeBSD ports, and Debian/Ubuntu (for the CI), and probably CentOS too. So that give us these possibilities:-

|Library   | C Interface? | FreeBSD port?  | dpkg?  | RPM?   |
|:-------|:-----|:-----|:-----|:-----|
|picojson | No (C++) | No  | No | No |
|jansson | Yes | Yes | Yes | Yes |
|parson | Yes | No | No | No |
|json-c | Yes | Yes | Yes | Yes |
|libcjson | Yes | Yes | Yes | No |

It's a simplistic analysis, and could do with some checking, but that gives us `jansson` and `json-c` as possibilities. Both of these are permissive licences.

I've not looked at either library yet. It's worth noting however than on a CentOS VM I've just run up for something else, json-c is required by systemd. So I'd say that's a solid choice as it's not going anywhere at present. Looking elsewhere, it's the only option on Slackware, which tends to be the hardest distribution for us to support.

There's a couple of caveats about using JSON which have occurred to me since our conversation.
1) We've standardised on using the [xrdp transport](https://github.com/neutrinolabs/xrdp/blob/devel/common/trans.h) mechanism for all connections (see also #1925). This can easily send json, but will find it harder to read it as normally messages have a binary header which contains the total length.
2) JSON is verbose compared to binary. We need to keep individual messages relatively small so they can be written without buffering at the application level, and also so we can easily solve 1).
3) We wouldn't be able to pass file descriptors around, or at least not embedded in the JSON. In the latest chat I had with @aquesnel about this, we decided that was OK, but things could change in the future.

I don't think any of these is a show-stopper. Performance isn't an issue for this interface.

I think the important thing is to make sure at the application level, this is a wrapped interface, at least initially so that if we've made a wrong decision it's easy to back it out. Later on, when the interfaces are stable, we can publish them and maybe move the utilities to using a scripting language with JSON support.

**Reply by metalefty** [reply]
Thanks for the quick analysis & report. I was also thinking jansson is a plausible option. Will analyse which is better jansson vs json-c.

**Reply by matt335672** [reply]
I have no idea which would suit us better at the moment, so I'm happy to go with whatever you think would work best.

A couple of other things:-
1. I've had a more thorough look, and although either package is available on all platforms we're likely to be running on, jansson is not available for the current release of Slackware 14.2. This will fixed when Slackware 15 is released, which should be soon (it's at RC1 at the moment). So there's really nothing to choose between them in terms of availability.
2. There's a fair bit to do to move to JSON for this interface. For example, we need to decide whether there's any benefit to using [JSON-RPC](https://en.wikipedia.org/wiki/JSON-RPC), or something similar. I'm still not certain on the best way to integrate it with the xrdp transport facility.

My current plan for SCP is first to get it all in one place with a single interface (I'm working on that now), and then produce an SCP V2 which supports all the features of V0 and V1 and lets us implement #1961. V0 and V1 will be dropped at this point. V2 will be a lot easier to use, but will still be a binary interface.

**Reply by metalefty** [reply]
Implementing SCP V2 first is a good idea.

I'm going to implement some proprietary features for our company's product that uses JSON-RPC. It is not related to sesman & authentication but I recalled you when I was thinking of JSON.  Now I'm inclined to use jansson 60%, json-c: 40%.

**Reply by matt335672** [reply]
Neither project appears to be particularly big, but then they probably don't need to be.

FWIW, libjansson appears to be the more popular package on my 20.04 development system, but there's not much in it:-

```
$ apt-cache rdepends libjansson4 | wc -l
90
$ apt-cache rdepends libjson-c4 | wc -l
71
```

**Comment by metalefty** [comment]
My company is making a proprietary extension of xrdp that uses JSON. Recently, I mostly spend time on that.  In the extension, I decided to Jansson so I'd appreciate it if you choose Jansson when mainline xrdp goes to SCPv2.
